File: lib/gui/main_window.py
# pylint: disable=E0611, E0401
from PyQt5.QtCore import Qt
from PyQt5.QtGui import QColor, QIcon, QPixmap
from PyQt5.QtWidgets import (
    QColorDialog,
    QComboBox,
    QFormLayout,
    QGroupBox,
    QLabel,
    QMainWindow,
    QPushButton,
    QVBoxLayout,
    QWidget,
)

from lib.models.battery_threshold import BatteryThreshold
from lib.models.boost import Boost
from lib.models.rgb_brightness import RgbBrightness
from lib.models.thermal_throttle_profile import ThermalThrottleProfile
from lib.services.openrgb_service import open_rgb_service
from lib.services.platform_service import platform_service
from lib.utils.constants import icons_path
from lib.utils.event_bus import event_bus
from lib.utils.gui_utils import GuiUtils
from lib.utils.logger import Logger
from lib.utils.translator import translator


class MainWindow(QMainWindow):
    """Aplication main window"""

    def __init__(self):
        self.logger = Logger()

        self.effect_labels = open_rgb_service.get_available_effects()

        super().__init__()

        self.setWindowTitle("RogControlCenter")
        self.setGeometry(0, 0, 580, 800)
        self.setFixedSize(580, 800)
        self.setWindowIcon(QIcon(f"{icons_path}/icon-45x45.png"))

        self.current_color = open_rgb_service.get_color(open_rgb_service.effect)

        # Widget central
        central_widget = QWidget(self)
        main_layout = QVBoxLayout(central_widget)
        main_layout.setContentsMargins(20, 20, 20, 20)
        main_layout.setAlignment(Qt.AlignTop)  # Alinear los elementos al inicio

        # Primera fila: Imagen centrada
        image_label = QLabel()
        pixmap = QPixmap(f"{icons_path}/rog-logo.svg")  # Ruta a tu archivo de imagen
        scaled_pixmap = pixmap.scaled(350, 350, Qt.KeepAspectRatio, Qt.SmoothTransformation)  # Escalar la imagen
        image_label.setPixmap(scaled_pixmap)
        image_label.setAlignment(Qt.AlignCenter)

        main_layout.addWidget(image_label, alignment=Qt.AlignCenter)  # Centrar la imagen

        # Grupo 1: Rendimiento
        performance_group = QGroupBox(translator.translate("performance"))
        performance_layout = QFormLayout()
        performance_layout.setContentsMargins(20, 20, 20, 20)

        # Label y Dropdown para "Perfil"
        self.profile_dropdown = QComboBox()
        for item in ThermalThrottleProfile:
            self.profile_dropdown.addItem(translator.translate(f"label.profile.{item.name}"), item)
        self.profile_dropdown.currentIndexChanged.connect(self.on_profile_changed)
        self.set_thermal_throttle_policy(platform_service.get_thermal_throttle_profile())
        performance_layout.addRow(QLabel(f"{translator.translate('profile')}:"), self.profile_dropdown)

        # Label y Dropdown para "Boost"
        self.boost_dropdown = QComboBox()
        for item in Boost:
            self.boost_dropdown.addItem(translator.translate(f"label.boost.{item.name}"), item)
        self.boost_dropdown.currentIndexChanged.connect(self.on_boost_changed)
        self.set_boost_mode(platform_service.get_boost_mode())
        performance_layout.addRow(QLabel(f"{translator.translate('boost')}:"), self.boost_dropdown)

        performance_group.setLayout(performance_layout)
        main_layout.addWidget(performance_group)

        # Grupo 2: Aura
        aura_group = QGroupBox("Aura")
        aura_layout = QFormLayout()
        aura_layout.setContentsMargins(20, 20, 20, 20)

        # Dropdown para "Efecto"
        self.effect_dropdown = QComboBox()
        self.effect_dropdown.addItems(self.effect_labels)
        self.effect_dropdown.setCurrentIndex(self.effect_labels.index(open_rgb_service.effect))
        self.effect_dropdown.currentIndexChanged.connect(self.on_effect_change)
        aura_layout.addRow(QLabel(f"{translator.translate("effect")}:"), self.effect_dropdown)

        # Dropdown para "Brillo"
        self.brightness_dropdown = QComboBox()
        for brightness in RgbBrightness:
            self.brightness_dropdown.addItem(translator.translate(f"label.brightness.{brightness.name}"), brightness)
        self.brightness_dropdown.setCurrentIndex(self.brightness_dropdown.findData(open_rgb_service.brightness))
        self.brightness_dropdown.currentIndexChanged.connect(self.on_brightness_change)
        aura_layout.addRow(QLabel(f"{translator.translate("brightness")}:"), self.brightness_dropdown)

        # Botón de color
        self.color_button = QPushButton()
        self.color_button.setFixedSize(50, 50)
        self.color_button.setStyleSheet(
            f"background-color: {self.current_color if self.current_color is not None else "#00000000"};"
        )
        self.color_button.clicked.connect(self.pick_color)
        supports_color = open_rgb_service.supports_color(open_rgb_service.effect)
        self.color_button.setDisabled(not supports_color)
        self.color_button.setHidden(not supports_color)

        self.color_label = QLabel(f"{translator.translate('color')}:")
        self.color_label.setHidden(not supports_color)
        aura_layout.addRow(self.color_label, self.color_button)

        aura_group.setLayout(aura_layout)
        main_layout.addWidget(aura_group)

        # Grupo 3: Configuración
        settings_group = QGroupBox(translator.translate("battery"))
        settings_layout = QFormLayout()
        settings_layout.setContentsMargins(20, 20, 20, 20)

        # Dropdown para "Umbral"
        self.threshold_dropdown = QComboBox()
        for item in BatteryThreshold:
            self.threshold_dropdown.addItem(f"{item.value}%", item)
        self.threshold_dropdown.currentIndexChanged.connect(self.on_battery_limit_changed)
        self.set_battery_charge_limit(platform_service.get_battery_charge_limit())
        settings_layout.addRow(
            QLabel(
                f"{translator.translate("charge.threshold")}:",
            ),
            self.threshold_dropdown,
        )

        settings_group.setLayout(settings_layout)
        main_layout.addWidget(settings_group)

        # Configurar el widget central
        self.setCentralWidget(central_widget)
        GuiUtils.center_window_on_current_screen(self)

        event_bus.on("PlatformService.battery_threshold", self.set_battery_charge_limit)
        event_bus.on("PlatformService.boost", self.set_boost_mode)
        event_bus.on("PlatformService.thermal_throttle_profile", self.set_thermal_throttle_policy)
        event_bus.on("OpenRgbService.aura_changed", self.set_aura_state)

    def closeEvent(self, event):  # pylint: disable=C0103
        """Override the close event to hide the window instead of closing it."""
        event.ignore()
        self.hide()

    def pick_color(self):
        """Open color picker"""
        color = QColorDialog.getColor(QColor(self.current_color), self, translator.translate("color.select"))
        if color.isValid():
            self.current_color = color.name().upper()
            self.color_button.setStyleSheet(f"background-color: {self.current_color};")
            self.on_color_change()

    def set_thermal_throttle_policy(self, value: ThermalThrottleProfile):
        """Set profile policy"""
        self.profile_dropdown.setCurrentIndex(self.profile_dropdown.findData(value))

    def set_boost_mode(self, value: Boost):
        """Set boost mode"""
        self.boost_dropdown.setCurrentIndex(self.boost_dropdown.findData(value))

    def set_battery_charge_limit(self, value: BatteryThreshold):
        """Set battery limit"""
        self.threshold_dropdown.setCurrentIndex(self.threshold_dropdown.findData(value))

    def set_aura_state(self, value):
        """Set aura state"""
        effect, brightness, color = value

        self.effect_dropdown.setCurrentIndex(self.effect_labels.index(effect))
        self.brightness_dropdown.setCurrentIndex(self.brightness_dropdown.findData(brightness))
        self.current_color = color if color is not None else "#00000000"
        self.color_button.setStyleSheet(f"background-color: {self.current_color};")
        self.color_button.setDisabled(color is None)
        self.color_button.setHidden(color is None)
        self.color_label.setHidden(color is None)

    def on_profile_changed(self, index):
        """Handler for profile change"""
        platform_service.set_thermal_throttle_policy(self.profile_dropdown.itemData(index))

    def on_boost_changed(self, index):
        """Handler for boost change"""
        platform_service.set_boost_mode(self.boost_dropdown.itemData(index))

    def on_battery_limit_changed(self, index):
        """Handler for battery limit change"""
        platform_service.set_battery_threshold(self.threshold_dropdown.itemData(index))

    def on_effect_change(self):
        """Handler for effect change"""
        effect = self.effect_labels[self.effect_dropdown.currentIndex()]
        open_rgb_service.apply_effect(effect)

    def on_brightness_change(self, index):
        """Handler for brightness change"""
        open_rgb_service.apply_brightness(self.brightness_dropdown.itemData(index))

    def on_color_change(self):
        """Handler for color change"""
        open_rgb_service.apply_color(self.current_color)


main_window = MainWindow()

File: RogPerfTuner/src/gui/main_window.cpp
#include "gui/main_window.hpp"

#include <qcombobox.h>
#include <qcontainerfwd.h>
#include <qlabel.h>
#include <qpushbutton.h>

#include <QApplication>
#include <QColorDialog>
#include <QFormLayout>
#include <QGroupBox>
#include <QLabel>
#include <QStatusBar>
#include <QVBoxLayout>
#include <algorithm>
#include <optional>

#include "OpenRGB/Color.hpp"
#include "framework/utils/string_utils.hpp"
#include "gui/fan_curve_editor.hpp"
#include "gui/game_list.hpp"
#include "models/performance/performance_profile.hpp"

MainWindow::MainWindow(QWidget* parent) : QMainWindow(parent) {
	onBattery	 = uPowerClient.isOnBattery();
	runningGames = steamService.getRunningGames().size();

	setWindowTitle((Constants::APP_NAME + " | " + Constants::APP_VERSION).c_str());
	setWindowIcon(QIcon(Constants::ASSET_ICON_45_FILE.c_str()));

	QWidget* centralWidget	= new QWidget(this);
	QVBoxLayout* mainLayout = new QVBoxLayout(centralWidget);
	mainLayout->setContentsMargins(20, 10, 20, 10);
	mainLayout->setAlignment(Qt::AlignTop);

	// -------------------------
	// Logo
	// -------------------------
	QLabel* imageLabel = new QLabel();
	QPixmap pixmap(Constants::ASSET_ICON_FILE.c_str());
	QPixmap scaledPixmap = pixmap.scaled(140, 140, Qt::KeepAspectRatio, Qt::SmoothTransformation);
	imageLabel->setPixmap(scaledPixmap);
	imageLabel->setAlignment(Qt::AlignCenter);
	mainLayout->addWidget(imageLabel, 0, Qt::AlignCenter);
	// -------------------------
	// Logo
	// -------------------------

	// -------------------------
	// Performance group
	// -------------------------
	QGroupBox* performanceGroup	   = new QGroupBox(translator.translate("performance").c_str());
	QFormLayout* performanceLayout = new QFormLayout();
	performanceLayout->setContentsMargins(20, 10, 20, 10);
	// -------------------------
	// Profile menu
	// -------------------------
	_profileDropdown = new NoScrollComboBox();
	_profileDropdown->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Preferred);
	auto items = values<PerformanceProfile>();
	for (PerformanceProfile item : items) {
		_profileDropdown->addItem(("  " + translator.translate("label.profile." + toName(item))).c_str(), toString(item).c_str());
	}
	connect(_profileDropdown, QOverload<int>::of(&QComboBox::currentIndexChanged), this, &MainWindow::onProfileChanged);
	setPerformanceProfile(performanceService.getPerformanceProfile());
	_profileDropdown->setEnabled(runningGames == 0 && !onBattery);
	performanceLayout->addRow(new QLabel((translator.translate("profile") + ":").c_str()), _profileDropdown);
	// -------------------------
	// Profile menu
	// -------------------------
	// -------------------------
	// Scheduler menu
	// -------------------------
	if (!performanceService.getAvailableSchedulers().empty()) {
		_schedulerDropdown = new NoScrollComboBox();
		_schedulerDropdown->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Preferred);
		_schedulerDropdown->addItem(("  " + performanceService.getDefaultSchedulerName()).c_str(), "");

		auto items2 = performanceService.getAvailableSchedulers();
		std::reverse(items2.begin(), items2.end());
		for (auto item : items2) {
			_schedulerDropdown->addItem(("  " + StringUtils::capitalize(item)).c_str(), item.c_str());
		}
		connect(_schedulerDropdown, QOverload<int>::of(&QComboBox::currentIndexChanged), this, &MainWindow::onSchedulerChanged);
		setScheduler(performanceService.getCurrentScheduler());
		_schedulerDropdown->setEnabled(runningGames == 0);
		performanceLayout->addRow(new QLabel((translator.translate("scheduler") + ":").c_str()), _schedulerDropdown);

		eventBus.onScheduler([this](std::optional<std::string> sched) {
			setScheduler(sched);
		});
	}
	// -------------------------
	// Scheduler menu
	// -------------------------
	// -------------------------
	// Fan curves menu
	// -------------------------
	if (!performanceService.getFans().empty()) {
		fanEdit = new QPushButton();
		fanEdit->setText(translator.translate("edit.curve").c_str());
		connect(fanEdit, &QPushButton::clicked, this, &MainWindow::openFanEditor);
		fanEdit->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Preferred);
		fanEdit->setEnabled(performanceService.getPerformanceProfile() != PerformanceProfile::SMART);

		performanceLayout->addRow(new QLabel((translator.translate("fan.curves") + ":").c_str()), fanEdit);
	}
	// -------------------------
	// Fan curves menu
	// -------------------------
	// -------------------------
	// Games menu
	// -------------------------
	_gameProfileButton = new QPushButton(translator.translate("label.game.configure").c_str());
	_gameProfileButton->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Preferred);
	connect(_gameProfileButton, &QPushButton::clicked, this, &MainWindow::openGameList);
	performanceLayout->addRow(new QLabel((translator.translate("games") + ":").c_str()), _gameProfileButton);
	// -------------------------
	// Games menu
	// -------------------------
	performanceGroup->setLayout(performanceLayout);
	mainLayout->addWidget(performanceGroup);

	eventBus.onPerformanceProfile([this](PerformanceProfile profile) {
		setPerformanceProfile(profile);
		fanEdit->setEnabled(profile != PerformanceProfile::SMART);
	});
	// -------------------------
	// Performance group
	// -------------------------

	// -------------------------
	// Aura group
	// -------------------------
	QGroupBox* auraGroup	= new QGroupBox("Aura");
	QFormLayout* auraLayout = new QFormLayout();
	auraLayout->setContentsMargins(20, 10, 20, 10);
	// -------------------------
	// Effect menu
	// -------------------------
	_effectDropdown = new NoScrollComboBox();
	_effectDropdown->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Preferred);
	auto effects = openRgbService.getAvailableEffects();
	for (const auto effect : effects) {
		_effectDropdown->addItem(("  " + effect).c_str(), effect.c_str());
	}

	auto it = std::find(effects.begin(), effects.end(), openRgbService.getCurrentEffect());
	_effectDropdown->setCurrentIndex(std::distance(effects.begin(), it));
	connect(_effectDropdown, QOverload<int>::of(&QComboBox::currentIndexChanged), this, &MainWindow::onEffectChange);
	auraLayout->addRow(new QLabel((translator.translate("effect") + ":").c_str()), _effectDropdown);

	eventBus.onRgbEffect([this](std::string effect) {
		setAuraEffect(effect);
	});
	// -------------------------
	// Effect menu
	// -------------------------
	// -------------------------
	// Brightness + Color menu
	// -------------------------
	_brightnessDropdown = new NoScrollComboBox();
	_brightnessDropdown->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Preferred);
	for (RgbBrightness b : values<RgbBrightness, true>()) {
		_brightnessDropdown->addItem(("  " + translator.translate("label.brightness." + toName(b))).c_str(), toInt(b));
	}
	_brightnessDropdown->setCurrentIndex(_brightnessDropdown->findData(toInt(openRgbService.getCurrentBrightness())));
	connect(_brightnessDropdown, QOverload<int>::of(&QComboBox::currentIndexChanged), this, &MainWindow::onBrightnessChange);

	_colorButton = new QPushButton();
	_colorButton->setFixedSize(25, 25);
	_colorButton->setStyleSheet(("background-color: " + (openRgbService.getColor().value_or(Color::Black.toHex()))).c_str());
	_colorButton->setCursor(Qt::PointingHandCursor);
	_colorButton->setEnabled(openRgbService.getColor().has_value());
	connect(_colorButton, &QPushButton::clicked, this, &MainWindow::showColorPicker);

	auto brightnessColorLayout = new QHBoxLayout();
	brightnessColorLayout->addWidget(_brightnessDropdown);
	brightnessColorLayout->addWidget(_colorButton);
	brightnessColorLayout->setSpacing(10);

	auraLayout->addRow(new QLabel((translator.translate("brightness") + ":").c_str()), brightnessColorLayout);

	auraGroup->setLayout(auraLayout);
	mainLayout->addWidget(auraGroup);

	eventBus.onRgbBrightness([this](RgbBrightness brightness) {
		setAuraBrightness(brightness);
	});

	eventBus.onRgbColor([this](std::optional<std::string> color) {
		setAuraColor(color);
	});
	// -------------------------
	// Brightness + Color menu
	// -------------------------
	// -------------------------
	// Aura group
	// -------------------------

	// -------------------------
	// Hardware group
	// -------------------------
	if (hardwareService.getBatteryLimitAvailable() || hardwareService.getBootSoundAvailable()) {
		QGroupBox* hardwareGroup	= new QGroupBox(translator.translate("hardware").c_str());
		QFormLayout* hardwareLayout = new QFormLayout();
		hardwareLayout->setContentsMargins(20, 10, 20, 10);
		// -------------------------
		// Battery menu
		// -------------------------
		if (hardwareService.getBatteryLimitAvailable()) {
			_thresholdDropdown = new NoScrollComboBox();
			_thresholdDropdown->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Preferred);
			for (BatteryThreshold t : values<BatteryThreshold, true>()) {
				const auto intVal = toInt(t);
				_thresholdDropdown->addItem("  " + QString::number(intVal) + "%", intVal);
			}
			connect(_thresholdDropdown, QOverload<int>::of(&QComboBox::currentIndexChanged), this, &MainWindow::onBatteryLimitChanged);
			setBatteryChargeLimit(hardwareService.getChargeThreshold());
			hardwareLayout->addRow(new QLabel((translator.translate("charge.threshold") + ":").c_str()), _thresholdDropdown);

			eventBus.onChargeThreshold([this](BatteryThreshold threshold) {
				setBatteryChargeLimit(threshold);
			});
		}
		// -------------------------
		// Battery menu
		// -------------------------
		// -------------------------
		// Boot sound menu
		// -------------------------
		if (hardwareService.getBootSoundAvailable()) {
			_bootSoundDropdown = new NoScrollComboBox();
			_bootSoundDropdown->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Preferred);

			auto vals = std::array<bool, 2>{true, false};
			for (size_t i = 0; i < vals.size(); i++) {
				_bootSoundDropdown->addItem(("  " + translator.translate(vals[i] ? "enabled" : "disabled")).c_str(), vals[i]);
			}
			onBootSoundEvent(hardwareService.getBootSound());

			connect(_bootSoundDropdown, QOverload<int>::of(&QComboBox::currentIndexChanged), this, &MainWindow::onBootSoundChanged);
			hardwareLayout->addRow(new QLabel(translator.translate("boot.sound").c_str()), _bootSoundDropdown);

			eventBus.onBootSound([this](bool value) {
				onBootSoundEvent(value);
			});
		}
		// -------------------------
		// Boot sound menu
		// -------------------------
		hardwareGroup->setLayout(hardwareLayout);
		mainLayout->addWidget(hardwareGroup);
	}
	// -------------------------
	// Hardware group
	// -------------------------

	// -------------------------
	// Settings group
	// -------------------------
	QGroupBox* settingsGroup	= new QGroupBox(translator.translate("settings").c_str());
	QFormLayout* settingsLayout = new QFormLayout();
	settingsLayout->setContentsMargins(20, 10, 20, 10);
	// -------------------------
	// Start minimized
	// -------------------------
	_minimized = new QCheckBox();
	_minimized->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Preferred);
	_minimized->setEnabled(true);
	_minimized->setChecked(applicationService.isStartMinimized());
	connect(_minimized, &QCheckBox::toggled, this, &MainWindow::onStartMinimizedChanged);
	settingsLayout->addRow(_minimized, new QLabel(translator.translate("start.minimized").c_str()));
	// -------------------------
	// Start minimized
	// -------------------------
	// -------------------------
	// Autostart
	// -------------------------
	_autostart = new QCheckBox();
	_autostart->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Preferred);
#ifdef DEV_MODE
	_autostart->setEnabled(false);
#endif
	_autostart->setChecked(applicationService.isAutostartEnabled());
	connect(_autostart, &QCheckBox::toggled, this, &MainWindow::onAutostartChanged);
	settingsLayout->addRow(_autostart, new QLabel(translator.translate("autostart").c_str()));
	// -------------------------
	// Autostart
	// -------------------------
	settingsGroup->setLayout(settingsLayout);
	mainLayout->addWidget(settingsGroup);
	// -------------------------
	// Settings group
	// -------------------------

	setCentralWidget(centralWidget);

	adjustSize();
	setFixedSize(size());

	QScreen* screen = QApplication::screenAt(QCursor::pos());
	if (screen) {
		QRect screenGeometry = screen->availableGeometry();
		int x				 = (screenGeometry.width() - width()) / 2 + screenGeometry.x();
		int y				 = (screenGeometry.height() - height()) / 2 + screenGeometry.y();
		move(x, y);
	}

	eventBus.onBattery([this](bool onBattery) {
		this->onBattery = onBattery;
		onBatteryEvent();
	});

	eventBus.onGameEvent([this](size_t runningGames) {
		this->runningGames = runningGames;
		onGameEvent();
	});
}

void MainWindow::closeEvent(QCloseEvent* event) {
	event->ignore();
	this->hide();
}

void MainWindow::onGameEvent() {
	_profileDropdown->setEnabled(runningGames == 0 && !onBattery);
	_schedulerDropdown->setEnabled(runningGames == 0);
}

void MainWindow::onBatteryEvent() {
	_profileDropdown->setEnabled(runningGames == 0 && !onBattery);
}

void MainWindow::onBootSoundEvent(bool value) {
	_bootSoundDropdown->setCurrentIndex(_bootSoundDropdown->findData(value));
}

void MainWindow::setPerformanceProfile(PerformanceProfile value) {
	_profileDropdown->setCurrentIndex(_profileDropdown->findData(toString(value).c_str()));
}

void MainWindow::setScheduler(std::optional<std::string> sched) {
	_schedulerDropdown->setCurrentIndex(_schedulerDropdown->findData(sched.value_or("").c_str()));
}

void MainWindow::setBatteryChargeLimit(BatteryThreshold value) {
	_thresholdDropdown->setCurrentIndex(_thresholdDropdown->findData(toInt(value)));
}

void MainWindow::setAuraBrightness(RgbBrightness brightness) {
	_brightnessDropdown->setCurrentIndex(_brightnessDropdown->findData(toInt(brightness)));
}

void MainWindow::setAuraColor(std::optional<std::string> color) {
	_colorButton->setStyleSheet(("background-color: " + (color.value_or(Color::Black.toHex()))).c_str());
	_colorButton->setEnabled(color.has_value());
}

void MainWindow::setAuraEffect(std::string effect) {
	_effectDropdown->setCurrentIndex(_effectDropdown->findData(effect.c_str()));
}

void MainWindow::onProfileChanged(int) {
	PerformanceProfile profile = fromString<PerformanceProfile>(_profileDropdown->currentData().toString().toStdString());
	if (steamService.getRunningGames().empty()) {
		if (performanceService.getPerformanceProfile() != profile) {
			performanceService.setPerformanceProfile(profile);
		}
	}
}

void MainWindow::onSchedulerChanged(int) {
	std::optional<std::string> scheduler = (_schedulerDropdown->currentData().toString().toStdString());
	if (scheduler.value_or("").empty()) {
		scheduler = std::nullopt;
	}

	if (scheduler != performanceService.getCurrentScheduler() && steamService.getRunningGames().empty()) {
		performanceService.setScheduler(scheduler);
	}
}

void MainWindow::onBatteryLimitChanged(int) {
	auto threshold = fromInt<BatteryThreshold>(_thresholdDropdown->currentData().toInt());
	if (hardwareService.getChargeThreshold() != threshold) {
		hardwareService.setChargeThreshold(threshold);
	}
}

void MainWindow::showColorPicker() {
	QColor initial = QColor(openRgbService.getColor().value().c_str());
	QColor chosen  = QColorDialog::getColor(initial, this, translator.translate("pick.color").c_str());

	if (chosen.isValid()) {
		openRgbService.setColor(chosen.name(QColor::HexRgb).toStdString());
	}
}

void MainWindow::onEffectChange() {
	std::string effect = openRgbService.getAvailableEffects()[_effectDropdown->currentIndex()];
	if (openRgbService.getCurrentEffect() != effect) {
		openRgbService.setEffect(effect);
	}
}

void MainWindow::onBrightnessChange(int) {
	auto level = fromInt<RgbBrightness>(_brightnessDropdown->currentData().toInt());
	if (openRgbService.getCurrentBrightness() != level) {
		openRgbService.setBrightness(level);
	}
}

void MainWindow::openGameList() {
	GameList* list = new GameList(this);
	list->show();
}

void MainWindow::openFanEditor() {
	auto profile		= _profileDropdown->currentData().toString().toStdString();
	CurveEditor* editor = new CurveEditor(profile, this);
	editor->show();
}

void MainWindow::onAutostartChanged(bool enabled) {
	applicationService.setAutostart(enabled);
}

void MainWindow::onStartMinimizedChanged(bool enabled) {
	applicationService.setStartMinimized(enabled);
}

void MainWindow::onBootSoundChanged(int) {
	auto enabled = _bootSoundDropdown->currentData().toBool();
	if (hardwareService.getBootSound() != enabled) {
		hardwareService.setBootSound(enabled);
	}
}